### DISCUSSION TITLE
Deferrable HttpSensor does not move to Triggerer when using `response_check`

### DISCUSSION THREAD

**Discussion by bejota** [apache/airflow#39597] [General] [answered]
Hello,
I'm using HttpSensor to poll SLURM's REST API for job status. I'm using `deferrable` mode with `response_check`. The task only goes to the triggerer if I remove the `response_check`. 

Airflow Docker Image `apache/airflow:latest-python3.9`
Airflow Version `2.9.0`
apache-airflow-providers-http            `4.10.0`

Does not move to triggerer:
```
check_slurm_job = HttpSensor(
    task_id="check_slurm_job",
    http_conn_id="slurm_conn_id",
    method="get",
    endpoint="slurm/v0.0.40/job/999",
    response_check=lambda response: "COMPLETED" in response.text,
    poke_interval=10,
    timeout=60 * 60 * 5,
    deferrable=True,  # not deferring to triggerer ?
    dag=dag,
)
```
Moves to triggerer:
```
check_slurm_job = HttpSensor(
    task_id="check_slurm_job",
    http_conn_id="slurm_conn_id",
    method="get",
    endpoint="slurm/v0.0.40/job/999",
    # response_check=lambda response: "COMPLETED" in response.text,
    poke_interval=10,
    timeout=60 * 60 * 5,
    deferrable=True,  # not deferring to triggerer ?
    dag=dag,
)
```
In digging further, I see that HttpTrigger has no `response_check` method. I will attempt to add one, unless there is a better approach to this problem?

**Comment by miguelomp** [comment]
I was migrating some `PythonSensor` that use the `requests` package to `HttpSensor` to leverage the deferrable feature, and encountered this behavior as well.

While prototyping to test the concept, I managed to get it working with a less-than-ideal fix. I have a FastAPI app with an endpoint that raises a 404 error when the external condition is not met. Otherwise, it returns a simple string, resulting in a 200 status code. By implementing this fix and bypassing the response check in `HttpSensor`, the deferrable feature functions correctly.

I don't know if I'll use it but meanwhile, it could work.

**Comment by josh-fell** [comment, marked as answer]
In the source code, [if `response_check` is specified, the HttpTrigger is not engaged](https://github.com/apache/airflow/blob/835f28c8b9141d7c1b9c527a77ad8da7bb460d08/airflow/providers/http/sensors/http.py#L163).

However, I agree this seems less than ideal. The HttpTrigger should simply be waiting for some non-failing response and pass the response to the `execute_complete()` method in the sensor for evaluation (should `response_check` be configured for the task).

This seems like a reasonable change. I'll convert this to an issue. @bejota by all means, if you have an implementation in mind, you are more than welcome to do so. The more the merrier here in Airflow!

**Reply by bejota** [reply]
Thank you!
I would think HttpOperator should have the same behavior, although I'm new to Airflow and not quite certain of the distinction between the two.
